### PR TITLE
Add support for completions inside string literals

### LINKED ISSUE #606: Provide completion lists for string literals
Many JavaScript frameworks are heavily based on using string values to do many things, from getting and setting attributes via the property name in string form, to defining event handlers to handle browsing to particular URLs. TypeScript added specialized signatures in overloads as one means to help mitigate the pain associated with these types of APIs. Unfortunately for a number of patterns this is insufficient and not general enough. Rather than try to design a specific type system enhancement that could handle some or all of the many ways these frameworks use string literals we could try some simple tricks in Visual Studio to help out. Namely, when typing a string literal the editor should provide a completion list based on other string literals already used in your program. Some other tools already provide similar capabilities (ex Sublime Text).

Consider something like this in a Backbone View class:

```TypeScript
class TodoView extends Backbone.View<TodoModel> {
    initialize() {
        this.listenTo(this.model, 'change', () => this.render());
        this.listenTo(this.model, 'todoDescriptionChanged', () => this.render());
        this.listenTo(this.model, 'destory', () => this.remove());
    }
```

At another point I may attempt to manually trigger one of these events:

```ts
    updateTodo() {
        var desc = this.$input.val().trim();
        var todo = this.model;
        if (desc) {
            todo.save({ title: desc });
            // provide completion list inside this string literal showing 
            // todoDescriptionChanged to avoid hard to debug typos 
            // and speed up development
            todo.trigger('todoDescriptionChanged'); 
        }
     }
```

This should be a relatively cheap solution to a common set of problems that we've always hoped TypeScript can improve on over JavaScript. If you did accidentally make a typo in one use of the literal this could also help you discover that when seeing 2 completion list items that are similar but slightly different.

### PR DESCRIPTION
**Pull request by mhegazy**
Fixes #606

This supports completions in string literals in the following locations:
 1. argument to a call expression whose target parameter has a StringLiteralType
 ```ts
declare function f(a: "A", b: number): void;
declare function f(a: "B", b: number): void;
declare function f(a: "C", b: number): void;

f(" // A | B | C
```
 2. argument expression of an element access expression
 ```ts
var o = {
    "#invlaid identifier name": 1
};

o["| // #invlaid identifier name
 ```

 3. a string literal contextually typed by a string literal type
 ```ts
type Options = "Option 1" | "Option 2" | "Option 3";
var x: Options = "| // Option 1 | Option 2 | Option 3
 ```

This PR does not cover:
* module name completion
* changing `o.` to `o[""]` if the completion entry is not a valid identifier name

### CODE REVIEW

File: src/services/services.ts [hunk 1]
```diff
@@ -4051,6 +4054,90 @@ namespace ts {
...
+                }
+                else if (isElementAccessExpression(node.parent) && node.parent.argumentExpression === node) {
+                    return getStringLiteralCompletionEntriesFromElementAccess(node.parent);
+                }
+                else {
+                    return getStringLiteralCompletionEntriesFromContextualType(<StringLiteral>node);
+                }
+            }
+
+            function getStringLiteralCompletionEntriesFromCallExpression(argumentInfo: SignatureHelp.ArgumentListInfo) {
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: I think we can eventually generalize this logic to other types as well

File: src/services/services.ts [hunk 2]
```diff
@@ -4051,6 +4054,90 @@ namespace ts {
...
+
+                for (const candidate of candidates) {
+                    if (candidate.parameters.length > argumentInfo.argumentIndex) {
+                        const parameter = candidate.parameters[argumentInfo.argumentIndex];
+                        addStringLiteralCompletionsFromType(typeChecker.getTypeAtLocation(parameter.valueDeclaration), entries);
+                    }
+                }
+
+                if (entries.length) {
+                    return { isMemberCompletion: false, isNewIdentifierLocation: true, entries: entries };
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: use shorthand

File: src/services/services.ts [hunk 3]
```diff
@@ -4051,6 +4054,90 @@ namespace ts {
...
+            }
+
+            function getStringLiteralCompletionEntriesFromElementAccess(node: ElementAccessExpression) {
+                const typeChecker = program.getTypeChecker();
+                const type = typeChecker.getTypeAtLocation(node.expression);
+                const entries: CompletionEntry[] = [];
+                if (type) {
+                    getCompletionEntriesFromSymbols(type.getApparentProperties(), entries, node, /*performCharacterChecks*/false);
+                    if (entries.length) {
+                        return { isMemberCompletion: true, isNewIdentifierLocation: true, entries: entries };
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: shorthand for `entries`

File: src/services/utilities.ts [hunk 4]
```diff
@@ -401,9 +401,27 @@ namespace ts {
...
+            // To be "in" one of these literals, the position has to be:
+            //   1. entirely within the token text.
+            //   2. at the end position of an unterminated token.
+            //   3. at the end of a regular expression (due to trailing flags like '/foo/g').
+            if (start < position && position < end) {
+                return true;
+            }
+
+            if (position === end) {
+                return !!(<LiteralExpression>previousToken).isUnterminated;
```
Comment (DanielRosenwasser) [inline src/services/utilities.ts]: i thought there was a `LiteralLike`

**Comment by DanielRosenwasser** [comment]
Very cool! Looks like this doesn't include the quote character, right?

File: src/services/services.ts [hunk 5]
```diff
@@ -4051,6 +4054,90 @@ namespace ts {
...
                 return uniqueNames;
             }
+
+            function getStringLiteralCompletionEntries(sourceFile: SourceFile, position: number) {
+                const node = findPrecedingToken(position, sourceFile);
+                if (!node || node.kind !== SyntaxKind.StringLiteral) {
+                    return undefined;
+                }
+
+                const argumentInfo = SignatureHelp.getContainingArgumentInfo(node, position, sourceFile);
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: This seems like a slightly weird sort of dependency. Ideally there would be an API on the type checker like `getPossibleContextualTypes`

File: src/services/signatureHelp.ts [hunk 6]
```diff
@@ -205,431 +205,431 @@ namespace ts.SignatureHelp {
...
-                    for (let declaration of declarations) {
-                        let symbol = declaration.symbol;
-                        if (symbol) {
-                            let type = typeChecker.getTypeOfSymbolAtLocation(symbol, declaration);
-                            if (type) {
-                                let callSignatures = type.getCallSignatures();
-                                if (callSignatures && callSignatures.length) {
-                                    return createSignatureHelpItems(callSignatures, callSignatures[0], argumentInfo);
-                                }
+        let typeChecker = program.getTypeChecker();
```
Comment (yuit) [inline src/services/signatureHelp.ts]: I think this should be `const`. We would want to add this files into linting maybe

File: src/services/services.ts (hunk 5 above)
Comment (mhegazy) [inline src/services/services.ts]: this is not contextual type per se. but sure. If there is another use case for such API we can push it into the checker, but this is the only one, i would keep it at the use site.

File: src/services/services.ts (hunk 1 above)
Comment (mhegazy) [inline src/services/services.ts]: i suppose if we do pre-selecting of completion entries based on contextual type.

File: src/services/signatureHelp.ts (hunk 6 above)
Comment (mhegazy) [inline src/services/signatureHelp.ts]: done.

**Comment by mhegazy** [comment]
> Very cool! Looks like this doesn't include the quote character, right?

yes. i am assuming here that automatic brace completion would get that done for you, or you have them written already. 

we should add replaceText, and replaceRange on CompletionEntry, to allow that. there are some issues with Roslyn API currently that makes this hard, but i believe should be better in next version.

**Comment by mhegazy** [comment]
@yuit and @DanielRosenwasser any more comments?

File: src/services/services.ts [hunk 7]
```diff
@@ -4051,6 +4054,90 @@ namespace ts {
...
+                if (argumentInfo) {
+                    return getStringLiteralCompletionEntriesFromCallExpression(argumentInfo);
+                }
+                else if (isElementAccessExpression(node.parent) && node.parent.argumentExpression === node) {
```
Comment (yuit) [inline src/services/services.ts]: Why you have to check this "node.parent.argumentExpression === node"?

File: src/services/services.ts [hunk 8]
```diff
@@ -4051,6 +4054,90 @@ namespace ts {
...
             }
+
+            function getStringLiteralCompletionEntries(sourceFile: SourceFile, position: number) {
+                const node = findPrecedingToken(position, sourceFile);
+                if (!node || node.kind !== SyntaxKind.StringLiteral) {
+                    return undefined;
+                }
+
+                const argumentInfo = SignatureHelp.getContainingArgumentInfo(node, position, sourceFile);
+                if (argumentInfo) {
```
Comment (yuit) [inline src/services/services.ts]: Minor: could you add comments from the PR description here. so we know what cases are covered

**Comment by yuit** [comment]
others than the small comments, look good :cake:.

File: src/services/services.ts [hunk 9]
```diff
@@ -3928,12 +3933,10 @@ namespace ts {
                 return { isMemberCompletion: false, isNewIdentifierLocation: false, entries: getAllJsDocCompletionEntries() };
             }
 
-            const sourceFile = getValidSourceFile(fileName);
-
             const entries: CompletionEntry[] = [];
 
             if (isSourceFileJavaScript(sourceFile)) {
-                const uniqueNames = getCompletionEntriesFromSymbols(symbols, entries);
```
Comment (yuit) [inline src/services/services.ts]: Why you start perform checking for characters?

File: src/services/services.ts (hunk 7 above)
Comment (mhegazy) [inline src/services/services.ts]: `"expression"["index"]` want to make sure we are on `"index"`

File: src/services/services.ts (hunk 8 above)
Comment (mhegazy) [inline src/services/services.ts]: sure.

File: src/services/services.ts (hunk 9 above)
Comment (mhegazy) [inline src/services/services.ts]: not sure really. this is how it was before adding the argument. let me look more into it.

File: src/services/signatureHelp.ts [hunk 10]
```diff
@@ -179,457 +179,457 @@ namespace ts.SignatureHelp {
...
-                                let callSignatures = type.getCallSignatures();
-                                if (callSignatures && callSignatures.length) {
-                                    return createSignatureHelpItems(callSignatures, callSignatures[0], argumentInfo);
-                                }
+        const typeChecker = program.getTypeChecker();
+        for (const sourceFile of program.getSourceFiles()) {
+            const nameToDeclarations = sourceFile.getNamedDeclarations();
+            const declarations = getProperty(nameToDeclarations, name.text);
+
+            if (declarations) {
```
Comment (vladima) [inline src/services/signatureHelp.ts]: nit: can you invert conditions to reduce nesting?

Comment (mhegazy) [inline src/services/signatureHelp.ts]: this is actually code that was there before, i just moved it around. i can get that cleaned in a future PR.

**Comment by Thaina** [comment]
Are there any jsdoc comment syntax I could made to use this feature in javascript with salsa?

**Comment by DanielRosenwasser** [comment]
Check out #9902 and #9995.

**Pull request merged**